### PR TITLE
Fixing team broadcast test tolerance

### PR DESCRIPTION
**Pull request by crtrott**
The test uses a floating point reduction and compares that to
another reduction result computed in a different order.
Needs a tolerance proportional to the number of reduction elements

Addresses issue #3526

### CODE REVIEW

File: core/unit_test/TestTeam.hpp [hunk 1]
```diff
@@ -1338,11 +1338,12 @@ struct TestTeamBroadcast<
...
   static inline
       typename std::enable_if<!std::is_integral<ScalarType>::value, void>::type
-      compare_test(ScalarType A, ScalarType B) {
-    if (std::is_same<ScalarType, double>::value) {
-      ASSERT_DOUBLE_EQ((double)A, (double)B);
-    } else if (std::is_same<ScalarType, float>::value) {
-      ASSERT_FLOAT_EQ((double)A, (double)B);
+      compare_test(ScalarType A, ScalarType B, double epsilon_factor) {
+    if (std::is_same<ScalarType, double>::value ||
+        std::is_same<ScalarType, float>::value) {
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:1343]: Any good reason not to use `std::is_floating_point`?

File: core/unit_test/TestTeam.hpp [hunk 2]
```diff
@@ -1338,11 +1338,12 @@ struct TestTeamBroadcast<
...
+      ASSERT_NEAR((double)A, (double)B,
+                  epsilon_factor * std::abs(A) *
+                      std::numeric_limits<ScalarType>::epsilon());
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:1346]: I am not a big fan of passing `epsilon_factor`, I'd pass directly the relative tolerance.

In any case, comment on what you are doing.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/unit_test/TestTeam.hpp (hunk 1 above)
Comment (crtrott) [reply core/unit_test/TestTeam.hpp:1343]: I was thinking of the addition of half and the ambiguity of what is_floating_point means in that case ...

File: core/unit_test/TestTeam.hpp (hunk 2 above)
Comment (crtrott) [reply core/unit_test/TestTeam.hpp:1346]: Isn't the comment below enough? The reason for the split is that I want to keep the float vs not float specifity etc. here in the comparison, while the test has to pass in the algorithmic epsilon. Hence my comment about the random walk.

Comment (dalg24) [reply core/unit_test/TestTeam.hpp:1346]: I can't really tell from the comment below nor the name of the function that this how you will use the trailing argument.  That said the comparison function is defined right above `TestTeamBroadcast` so it is not a big deal if you don't want to change.

**Review by masterleinad — APPROVED** [review]
Looks reasonable.

File: core/unit_test/TestTeam.hpp (hunk 2 above)
Comment (DavidPoliakoff) [reply core/unit_test/TestTeam.hpp:1346]: I think the deciding factor here is whether we think we could ever want a different epsilon_factor for different types, which I don't think we ever will.

File: core/unit_test/TestTeam.hpp (hunk 1 above)
Comment (mhoemmen) [reply core/unit_test/TestTeam.hpp:1343]: > I was thinking of the addition of `half` and the ambiguity of what `is_floating_point` means in that case ...

It means `float`, `double`, and `long double`.  Other types are not "floating-point types" in C++ <= 20.  If one of those papers that proposes generalizing this notion lands, then you might like to ask `is_iec559` and write generic IEEE 754 - flavored bounds.

**Pull request merged**